ClassFieldTheory/Mathlib/ValuativeLemmas.lean

```diff
@@ -0,0 +1,218 @@
+topology. This is used in `ValuativeTopology` for the statement that a valuative relation is
+compatible with a given topology -/
+def ball (v : Valuation R Γ) (γ : Γ) : Set R :=
+  { x | v x < γ }
+
+omit [ValuativeRel R] in
+@[simp] lemma mem_ball_iff (x : R) (γ : Γ) :
+    x ∈ v.ball γ ↔ v x < γ := by
+  rw [ball, Set.mem_setOf_eq]
+
+namespace Compatible
+
+lemma rel_one_iff :
+    x ≤ᵥ 1 ↔ v x ≤ 1 := by
+  rw [← map_one v, ← Valuation.Compatible.rel_iff_le]
+
+lemma rel_zero_iff :
+    x ≤ᵥ 0 ↔ v x ≤ 0 := by
+  rw [← map_zero v, ← Valuation.Compatible.rel_iff_le]
+
+lemma one_rel_iff :
+    1 ≤ᵥ x ↔ 1 ≤ v x := by
+  rw [← map_one v, ← Valuation.Compatible.rel_iff_le]
+
+section compare
+
+-- `v` is placed first here because that's the valuation one would typically want to rewrite into.
+variable (v' : Valuation R Γ') [Valuation.Compatible v']
+
+lemma lt_iff_lt (x y : R) : v' x < v' y ↔ v x < v y := by
+  simp_rw [lt_iff_le_not_ge, ← Valuation.Compatible.rel_iff_le]
+
+lemma ball_eq_ball (x : R) :
+    v'.ball (v' x) = v.ball (v x) := by
+  ext y; simp_rw [mem_ball_iff, Valuation.Compatible.lt_iff_lt v v']
+
+lemma lt_one_iff_lt_one (x : R) : v' x < 1 ↔ v x < 1 := by
+  rw [← v.map_one, ← v'.map_one, lt_iff_lt v v']
+
+lemma one_lt_iff_one_lt (x : R) : 1 < v' x ↔ 1 < v x := by
+  rw [← v.map_one, ← v'.map_one, lt_iff_lt v v']
+
+end compare
+
+end Valuation.Compatible
+
+
+namespace ValuativeRel
+
+@[ext] theorem ext {R : Type u} [CommRing R] {v v' : ValuativeRel R}
+    (h : ∀ a b, v.rel a b ↔ v'.rel a b) : v = v' := by
+  cases v; cases v'; congr; ext; apply h
+
+theorem rel_iff_div_rel_one {F : Type u} [Field F] [ValuativeRel F]
+    (x : F) {y : F} (hy : y ≠ 0) : x ≤ᵥ y ↔ x / y ≤ᵥ 1 := by
+  rw [Valuation.Compatible.rel_iff_le (v := ValuativeRel.valuation F),
+    Valuation.Compatible.rel_iff_le (v := ValuativeRel.valuation F),
+    map_div₀, map_one, div_le_one₀ (bot_lt_iff_ne_bot.2 ((map_ne_zero _).2 hy))]
+
+theorem rel_zero_iff {F : Type u} [Field F] [ValuativeRel F] (x : F) :
+    x ≤ᵥ 0 ↔ x = 0 := by
+  rw [Valuation.Compatible.rel_iff_le (v := valuation F), map_zero, le_zero_iff, map_eq_zero]
+
+/-- Two valuative relations on a field are equal iff their rings of integers are equal. -/
+@[ext high] theorem ext_field {F : Type u} [Field F] {v v' : ValuativeRel F}
+    (h : ∀ x, v.rel x 1 ↔ v'.rel x 1) : v = v' := by
+  ext x y
+  by_cases hy : y = 0
+  · simp_rw [hy, rel_zero_iff]
+  · rw [rel_iff_div_rel_one _ hy, @rel_iff_div_rel_one _ _ v x y hy, h]
+
+theorem posSubmonoid.ne_zero {R : Type u} [CommRing R] [ValuativeRel R]
+    (x : posSubmonoid R) : x.val ≠ 0 :=
+  mt (· ▸ rel_rfl) x.2
+
+theorem valuation_surjective₀ {F : Type u} [Field F] [ValuativeRel F]
+    (γ : ValueGroupWithZero F) : ∃ x : F, valuation F x = γ :=
+  let ⟨x, y, hxy⟩ := valuation_surjective γ
+  ⟨x / y.val, by rw [map_div₀, hxy]⟩
+
+theorem units_map_valuation_surjective {F : Type u} [Field F] [ValuativeRel F]
```

Comment:
```suggestion
theorem unitsMap_valuation_surjective {F : Type u} [Field F] [ValuativeRel F]
```

Reply:
And why is this not stated in terms of `Function.Surjective`?

Reply:
done